File: rewrite/src/main/java/com/ramitsuri/expensemanager/entities/Filter.java
package com.ramitsuri.expensemanager.entities;

import android.os.Parcel;
import android.os.Parcelable;
import android.util.Pair;
import android.util.SparseArray;

import com.ramitsuri.expensemanager.constants.Constants;
import com.ramitsuri.expensemanager.constants.intDefs.RecordType;
import com.ramitsuri.expensemanager.data.utils.SqlBuilder;
import com.ramitsuri.expensemanager.utils.AppHelper;
import com.ramitsuri.expensemanager.utils.DateHelper;
import com.ramitsuri.expensemanager.utils.ObjectHelper;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;
import java.util.TimeZone;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;

import androidx.sqlite.db.SimpleSQLiteQuery;

import timber.log.Timber;

public class Filter implements Parcelable {
    private SparseArray<Pair<Long, Long>> mDateTimes;
    private Boolean mIsIncome;
    private List<String> mCategories;
    private List<String> mPaymentMethods;
    private Boolean mIsSynced;
    private Boolean mIsStarred;
    @RecordType
    private String mRecordType;

    public Filter() {
    }

    protected Filter(Parcel in) {
        int[] months = in.createIntArray();
        if (months != null) {
            for (int monthIndex : months) {
                if (monthIndex == Constants.Basic.UNDEFINED) {
                    continue;
                }
                addMonthIndex(monthIndex);
            }
        }
        String isIncome = in.readString();
        if (isIncome != null) {
            mIsIncome = isIncome.equals(TRUE);
        }
        mCategories = in.createStringArrayList();
        mPaymentMethods = in.createStringArrayList();
        String isSynced = in.readString();
        if (isSynced != null) {
            mIsSynced = isSynced.equals(TRUE);
        }
        String isStarred = in.readString();
        if (isStarred != null) {
            mIsStarred = isStarred.equals(TRUE);
        }
        mRecordType = in.readString();
    }

    public static final Creator<Filter> CREATOR = new Creator<Filter>() {
        @Override
        public Filter createFromParcel(Parcel in) {
            return new Filter(in);
        }

        @Override
        public Filter[] newArray(int size) {
            return new Filter[size];
        }
    };

    public Filter getDefault() {
        return setIsIncome(false)
                .addCurrentMonth(AppHelper.getTimeZone());
    }

    public Filter clear() {
        mDateTimes = null;
        mIsIncome = null;
        mCategories = null;
        mPaymentMethods = null;
        mIsStarred = null;
        mIsSynced = null;
        mRecordType = RecordType.MONTHLY;

        return this;
    }

    public Filter addMonthIndex(int monthIndex) {
        return addMonthIndex(monthIndex, AppHelper.getTimeZone());
    }

    public Filter addMonthIndex(int monthIndex, TimeZone timeZone) {
        if (mDateTimes == null) {
            mDateTimes = new SparseArray<>();
        }
        onMonthAddRequested(monthIndex, timeZone);
        return this;
    }

    public Filter addCurrentMonth(TimeZone timeZone) {
        addMonthIndex(Constants.Basic.UNDEFINED, timeZone);
        return this;
    }

    public Filter removeMonthIndex(int monthIndex) {
        onMonthRemoveRequested(monthIndex);
        return this;
    }

    @Nullable
    public SparseArray<Pair<Long, Long>> getDateTimes() {
        return mDateTimes;
    }

    @Nullable
    public Boolean getIsIncome() {
        return mIsIncome;
    }

    public Filter setIsIncome(Boolean isIncome) {
        mIsIncome = isIncome;
        return this;
    }

    public List<String> getCategories() {
        return mCategories;
    }

    public Filter addCategory(String category) {
        if (mCategories == null) {
            mCategories = new ArrayList<>();
        }
        mCategories.add(category);
        return this;
    }

    public Filter removeCategory(String category) {
        if (mCategories != null) {
            mCategories.remove(category);
        }
        return this;
    }

    public List<String> getPaymentMethods() {
        return mPaymentMethods;
    }

    public Filter addPaymentMethod(String paymentMethod) {
        if (mPaymentMethods == null) {
            mPaymentMethods = new ArrayList<>();
        }
        mPaymentMethods.add(paymentMethod);
        return this;
    }

    public Filter removePaymentMethod(String paymentMethod) {
        if (mPaymentMethods != null) {
            mPaymentMethods.remove(paymentMethod);
        }
        return this;
    }

    public Boolean getIsSynced() {
        return mIsSynced;
    }

    public Filter setSynced(Boolean synced) {
        mIsSynced = synced;
        return this;
    }

    public Boolean getIsStarred() {
        return mIsStarred;
    }

    public Filter setIsStarred(Boolean starred) {
        mIsStarred = starred;
        return this;
    }

    @RecordType
    public String getRecordType() {
        return mRecordType;
    }

    public Filter setRecordType(@RecordType String recordType) {
        mRecordType = recordType;
        return this;
    }

    @Nonnull
    @Override
    public String toString() {
        return "Filter{" +
                "mDateTimes=" + mDateTimes +
                ", mIsIncome=" + mIsIncome +
                ", mCategories=" + mCategories +
                ", mPaymentMethods=" + mPaymentMethods +
                ", mIsSynced=" + mIsSynced +
                ", mIsStarred=" + mIsStarred +
                ", mRecordType=" + mRecordType +
                '}';
    }

    @Nullable
    public String toFriendlyString() {
        String friendlyString = null;
        if (mDateTimes != null && mDateTimes.size() == 1) {
            friendlyString = DateHelper.getMonth(mDateTimes.valueAt(0).first);
        }
        if (mIsSynced != null || mIsStarred != null ||
                (mCategories != null && mCategories.size() > 0) ||
                (mPaymentMethods != null && mPaymentMethods.size() > 0) ||
                (mRecordType == null || mRecordType.equals(RecordType.ANNUAL))) {
            return null;
        }
        return friendlyString;
    }

    private void onMonthAddRequested(int monthIndex, TimeZone timeZone) {
        Calendar cal = Calendar.getInstance(timeZone);
        if (monthIndex == Constants.Basic.UNDEFINED) {
            monthIndex = cal.get(Calendar.MONTH);
        }
        if (monthIndex < 0 || monthIndex > 11) {
            Timber.i("Invalid month index");
            return;
        }
        if (mDateTimes.get(monthIndex) != null) {
            Timber.i("Month already contained in filter");
            return;
        }

        cal.set(Calendar.MONTH, monthIndex);

        // First day of month at 00:00:00 001ms
        cal.set(Calendar.DAY_OF_MONTH, cal.getActualMinimum(Calendar.DAY_OF_MONTH));
        cal.set(Calendar.HOUR_OF_DAY, 0);
        cal.set(Calendar.MINUTE, 0);
        cal.set(Calendar.SECOND, 0);
        cal.set(Calendar.MILLISECOND, 0);
        long fromDateTime = cal.getTimeInMillis();

        // Last Day of month at 23:59:59 999ms
        cal.set(Calendar.DAY_OF_MONTH, cal.getActualMaximum(Calendar.DAY_OF_MONTH));
        cal.set(Calendar.HOUR_OF_DAY, 23);
        cal.set(Calendar.MINUTE, 59);
        cal.set(Calendar.SECOND, 59);
        cal.set(Calendar.MILLISECOND, 999);
        long toDateTime = cal.getTimeInMillis();
        Pair<Long, Long> dateTime = new Pair<>(fromDateTime, toDateTime);

        mDateTimes.put(monthIndex, dateTime);
    }

    private void onMonthRemoveRequested(int monthIndex) {
        if (mDateTimes == null) {
            Timber.i("mDateTimes null");
            return;
        }
        if (monthIndex < 0 || monthIndex > 11) {
            Timber.i("Invalid month index");
            return;
        }

        mDateTimes.remove(monthIndex);
    }

    public SimpleSQLiteQuery toQuery() {
        SqlBuilder builder = new SqlBuilder();
        builder.select("*")
                .from(Expense.DB.TABLE);
        // Is income
        if (getIsIncome() != null) {
            builder.where()
                    .column(Expense.DB.COL_INCOME)
                    .equal(getIsIncome() ? 1 : 0);
        }
        // Date ranges
        if (getDateTimes() != null && getDateTimes().size() > 0) {
            builder.whereOrAnd()
                    .between(Expense.DB.COL_DATE_TIME,
                            ObjectHelper.sparseArrayToList(getDateTimes()));
        }
        // Categories
        if (getCategories() != null && getCategories().size() > 0) {
            builder.whereOrAnd()
                    .column(Expense.DB.COL_CATEGORY)
                    .in(getCategories());
        }

        // Payment Methods
        if (getPaymentMethods() != null && getPaymentMethods().size() > 0) {
            builder.whereOrAnd()
                    .column(Expense.DB.COL_PAYMENT)
                    .in(getPaymentMethods());
        }

        // Synced
        if (getIsSynced() != null) {
            builder.whereOrAnd()
                    .column(Expense.DB.COL_SYNCED)
                    .equal(getIsSynced() ? 1 : 0);
        }

        // Starred
        if (getIsStarred() != null) {
            builder.whereOrAnd()
                    .column(Expense.DB.COL_STARRED)
                    .equal(getIsStarred() ? 1 : 0);
        }

        // Record Type
        if (getRecordType() != null) {
            builder.whereOrAnd()
                    .column(Expense.DB.COL_RECORD_TYPE)
                    .equal(getRecordType());
        }
        SimpleSQLiteQuery query =
                new SimpleSQLiteQuery(builder.toString(), builder.getArgs().toArray());
        Timber.i("Generated query is: [%s]", query.getSql());
        return query;
    }

    public SimpleSQLiteQuery toUpdateSyncedQuery() {
        SqlBuilder builder = new SqlBuilder();
        builder.update(Expense.DB.TABLE)
                .setEqualTo(Expense.DB.COL_SYNCED, 0);

        // Date ranges
        if (getDateTimes() != null && getDateTimes().size() > 0) {
            builder.whereOrAnd()
                    .between(Expense.DB.COL_DATE_TIME,
                            ObjectHelper.sparseArrayToList(getDateTimes()));
        }
        SimpleSQLiteQuery query =
                new SimpleSQLiteQuery(builder.toString(), builder.getArgs().toArray());
        Timber.i("Generated query is: [%s]", query.getSql());
        return query;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        if (mDateTimes != null) {
            int[] array = new int[mDateTimes.size()];
            for (int i = 0; i < mDateTimes.size(); i++) {
                if (mDateTimes.get(mDateTimes.keyAt(i)) == null) {
                    array[i] = Constants.Basic.UNDEFINED;
                } else {
                    array[i] = mDateTimes.keyAt(i);
                }
            }
            dest.writeIntArray(array);
        } else {
            dest.writeIntArray(null);
        }
        if (mIsIncome != null) {
            if (mIsIncome) {
                dest.writeString(TRUE);
            } else {
                dest.writeString(FALSE);
            }
        } else {
            dest.writeString(null);
        }
        if (mCategories != null) {
            dest.writeStringList(mCategories);
        } else {
            dest.writeStringList(null);
        }
        if (mPaymentMethods != null) {
            dest.writeStringList(mPaymentMethods);
        } else {
            dest.writeStringList(null);
        }
        if (mIsSynced != null) {
            if (mIsSynced) {
                dest.writeString(TRUE);
            } else {
                dest.writeString(FALSE);
            }
        } else {
            dest.writeString(null);
        }
        if (mIsStarred != null) {
            if (mIsStarred) {
                dest.writeString(TRUE);
            } else {
                dest.writeString(FALSE);
            }
        } else {
            dest.writeString(null);
        }
        if (mRecordType != null) {
            dest.writeString(mRecordType);
        } else {
            dest.writeString(null);
        }
    }

    private static final String TRUE = "true";
    private static final String FALSE = "false";
}

File: rewrite/src/main/java/com/ramitsuri/expensemanager/entities/Filter.kt
package com.ramitsuri.expensemanager.entities

import android.os.Parcel
import android.os.Parcelable
import androidx.sqlite.db.SimpleSQLiteQuery
import com.ramitsuri.expensemanager.constants.intDefs.RecordType
import com.ramitsuri.expensemanager.data.utils.SqlBuilder
import com.ramitsuri.expensemanager.utils.AppHelper
import com.ramitsuri.expensemanager.utils.DateHelper
import timber.log.Timber
import java.time.YearMonth
import java.time.ZoneId
import java.util.*
import kotlin.collections.ArrayList

class Filter(
        var years: MutableSet<Int>? = null,
        var months: MutableSet<Int>? = null,
        var categories: MutableList<String>? = null,
        var paymentMethods: MutableList<String>? = null,
        @RecordType
        var recordType: String? = null,
        var isStarred: Boolean? = null,
        private var timeZone: TimeZone = AppHelper.getTimeZone()
): Parcelable {
    constructor(parcel: Parcel): this() {
        val yearArray = parcel.createIntArray()
        yearArray?.let {
            years = it.toMutableSet()
        }

        val monthArray = parcel.createIntArray()
        monthArray?.let {
            months = it.toMutableSet()
        }

        categories = parcel.createStringArrayList()

        paymentMethods = parcel.createStringArrayList()

        recordType = parcel.readString()

        val starred: String? = parcel.readString()
        if (starred != null) {
            isStarred = starred == TRUE
        }

        timeZone = TimeZone.getTimeZone(ZoneId.of(parcel.readString()))
    }

    constructor(timeZone: TimeZone): this() {
        this.timeZone = timeZone
    }

    fun getDefault(): Filter {
        apply {
            addCurrentMonth()
        }
        return this
    }

    fun clear(): Filter {
        years = null
        months = null
        categories = null
        paymentMethods = null
        isStarred = null
        recordType = RecordType.MONTHLY
        return this
    }

    fun addYear(year: Int): Filter {
        if (years == null) {
            years = mutableSetOf()
        }
        years?.add(year)
        return this
    }

    fun removeYear(year: Int): Filter {
        years?.remove(year)
        return this
    }

    fun addMonth(month: Int): Filter {
        if (months == null) {
            months = mutableSetOf()
        }
        months?.add(month)
        return this
    }

    fun removeMonth(month: Int): Filter {
        months?.remove(month)
        return this
    }

    private fun addCurrentMonth(): Filter {
        val yearMonth = YearMonth.now(timeZone.toZoneId())
        addYear(yearMonth.year)
        addMonth(yearMonth.month.value)
        return this
    }

    fun addCategory(category: String): Filter {
        if (categories == null) {
            categories = ArrayList()
        }
        categories?.add(category)
        return this
    }

    fun removeCategory(category: String): Filter {
        categories?.remove(category)
        return this
    }

    fun addPaymentMethod(paymentMethod: String): Filter {
        if (paymentMethods == null) {
            paymentMethods = ArrayList()
        }
        paymentMethods?.add(paymentMethod)
        return this
    }

    fun removePaymentMethod(paymentMethod: String): Filter {
        paymentMethods?.remove(paymentMethod)
        return this
    }

    fun toFriendlyString(): String? {
        var friendlyString: String? = null
        val periods = getPeriods()
        if (periods.size == 1) {
            friendlyString = DateHelper.getMonthAndYear(periods[0].start, timeZone)
        }
        return if (isStarred != null ||
                categories != null && categories!!.size > 0 ||
                paymentMethods != null && paymentMethods!!.size > 0 ||
                recordType == null || recordType == RecordType.ANNUAL) {
            null
        } else friendlyString
    }

    fun toQuery(): SimpleSQLiteQuery {
        val builder = SqlBuilder()
        builder.select("*")
                .from(Expense.DB.TABLE)
        // Date ranges
        val periods = getPeriods()
        periods.let {
            if (it.isNotEmpty()) {
                builder.whereOrAnd()
                        .betweenPeriods(Expense.DB.COL_DATE_TIME, it)
            }
        }
        // Categories
        categories?.let {
            if (it.isNotEmpty()) {
                builder.whereOrAnd()
                        .column(Expense.DB.COL_CATEGORY)
                        .`in`<String>(it)
            }
        }
        // Payment Methods
        paymentMethods?.let {
            if (it.isNotEmpty()) {
                builder.whereOrAnd()
                        .column(Expense.DB.COL_PAYMENT)
                        .`in`<String>(it)
            }
        }
        // Starred
        isStarred?.let {
            builder.whereOrAnd()
                    .column(Expense.DB.COL_STARRED)
                    .equal(if (it) 1 else 0)
        }

        // Record Type
        recordType?.let {
            builder.whereOrAnd()
                    .column(Expense.DB.COL_RECORD_TYPE)
                    .equal(it)
        }
        val query = SimpleSQLiteQuery(builder.toString(), builder.args.toTypedArray())
        Timber.i("Generated query is: [%s]", query.sql)
        return query
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        var yearArray: IntArray? = null
        years?.let {
            yearArray = it.toIntArray()
        }
        dest.writeIntArray(yearArray)

        var monthArray: IntArray? = null
        months?.let {
            monthArray = it.toIntArray()
        }
        dest.writeIntArray(monthArray)

        categories?.let {
            dest.writeStringList(it)
        } ?: run {
            dest.writeStringList(null)
        }

        paymentMethods?.let {
            dest.writeStringList(it)
        } ?: run {
            dest.writeStringList(null)
        }

        recordType?.let {
            dest.writeString(it)
        } ?: run {
            dest.writeString(null)
        }

        isStarred?.let {
            if (it) {
                dest.writeString(TRUE)
            } else {
                dest.writeString(FALSE)
            }
        } ?: run {
            dest.writeString(null)
        }

        dest.writeString(timeZone.toZoneId().id)
    }

    fun getPeriods(): List<Period> {
        val periods = mutableListOf<Period>()
        years?.let {years ->
            months?.let {months ->
                for (year in years) {
                    for (month in months) {
                        periods.add(Period.fromYearAndMonth(year, month, timeZone))
                    }
                }
            }
        }
        return periods
    }

    companion object CREATOR: Parcelable.Creator<Filter> {
        override fun createFromParcel(parcel: Parcel): Filter {
            return Filter(parcel)
        }

        override fun newArray(size: Int): Array<Filter?> {
            return arrayOfNulls(size)
        }

        private const val TRUE = "true"
        private const val FALSE = "false"
    }
}